### PR TITLE
preliminary privlege escalation unification  + pbrun

### PR DESCRIPTION
**Pull request by bcoca**
TODO: add docs and update examples
- become constants inherit existing sudo/su ones
- become command line options, marked sudo/su as deprecated and moved sudo/su passwords to runas group
- changed method signatures as privlege escalation is collapsed to become
- added tests for su and become, diabled su for lack of support in local.py
- updated playbook,play and task objects to become
- added become to runner
- added whoami test for become/sudo/su
- added home override dir for plugins
- removed useless method from ask pass
- forced become pass to always be string also uses to_bytes
- fixed fakerunner for tests
- corrected reference in synchronize action plugin
- added pfexec (needs testing)
- removed unused sudo/su in runner init
- removed deprecated info
- updated pe tests to allow to run under sudo and not need root
- normalized become options into a funciton to avoid duplication and inconsistencies
- pushed suppored list to connection classs property
- updated all connection plugins to latest 'become' pe

### CODE REVIEW

File: lib/ansible/utils/__init__.py [hunk 1]
```diff
@@ -1025,22 +1023,35 @@ def base_parser(constants=C, usage="", output_opts=False, runas_opts=False,
...
+        parser.add_option('-S', '--su', default=constants.DEFAULT_SU, action='store_true',
+            help='run operations with su (deprecated, use become)')
+        parser.add_option('-R', '--su-user', default=None,
+            help='run operations with su as this user (default=%s) (deprecated, use become)' % constants.DEFAULT_SU_USER)
+
+        # consolidated privlege escalation (become)
+        parser.add_option("-b", "--become", default=constants.DEFAULT_BECOME, action="store_true", dest='become',
+            help="run operations with become (nopasswd implied)")
+        parser.add_option('--become-method', dest='become_method', default=constants.DEFAULT_BECOME_METHOD, type='string',
+            help="privlege escalation method to use (default=%s), valid choices: [ %s ]" % (constants.DEFAULT_BECOME_METHOD, ' | '.join(constants.BECOME_METHODS)))
```
Comment (mscherer) [inline lib/ansible/utils/__init__.py]: Typo in the help message

File: lib/ansible/utils/__init__.py [hunk 2]
```diff
@@ -1025,22 +1023,35 @@ def base_parser(constants=C, usage="", output_opts=False, runas_opts=False,
...
+        parser.add_option('--become-user', default=None, dest='become_user', type='string',
+            help='run operations as this user (default=%s)' % constants.DEFAULT_BECOME_USER)
+        parser.add_option('--ask-become-pass', default=False, dest='become_ask_pass', action='store_true',
+            help='ask for privlege escalation password')
```
Comment (mscherer) [inline lib/ansible/utils/__init__.py]: Typo in the help message

File: lib/ansible/utils/__init__.py (hunk 1 above)
Comment (bcoca) [inline lib/ansible/utils/__init__.py]: what typo? my tired eyes don't see it

File: lib/ansible/runner/connection_plugins/paramiko_ssh.py [hunk 3]
```diff
@@ -224,50 +230,46 @@ def exec_command(self, cmd, tmp_path, sudo_user=None, sudoable=False, executable
...
                     else:
-                        no_prompt_out += sudo_output
-                        no_prompt_err += sudo_output
+                        no_prompt_out += become_output
+                        no_prompt_err += become_output
 
             except socket.timeout:
 
-                raise errors.AnsibleError('ssh timed out waiting for sudo.\n' + sudo_output)
+                raise errors.AnsibleError('ssh timed out waiting for privelege escalation.\n' + become_output)
```
Comment (mscherer) [inline lib/ansible/runner/connection_plugins/paramiko_ssh.py]: Typo in privelege

Comment (bcoca) [inline lib/ansible/runner/connection_plugins/paramiko_ssh.py]: fixed

File: bin/ansible [hunk 4]
```diff
@@ -82,12 +82,8 @@ class Cli(object):
...
             sys.exit(1)
 
-        # su and sudo command line arguments need to be mutually exclusive
-        if (options.su or options.su_user or options.ask_su_pass) and \
-                (options.sudo or options.sudo_user or options.ask_sudo_pass):
-            parser.error("Sudo arguments ('--sudo', '--sudo-user', and '--ask-sudo-pass') "
-                         "and su arguments ('-su', '--su-user', and '--ask-su-pass') are "
-                         "mutually exclusive")
+        # privlege escalation command line arguments need to be mutually exclusive
+        utils.check_mutually_exclusive_privelege(options, parser)
```
Comment (mscherer) [inline bin/ansible]: Typo in the function name ( s/privelege/privilege/ )

File: lib/ansible/constants.py [hunk 5]
```diff
@@ -86,9 +86,6 @@ def shell_expand_path(path):
         path = os.path.expanduser(os.path.expandvars(path))
     return path
 
-def get_plugin_paths(path):
-    return ':'.join([os.path.join(x, path) for x in [os.path.expanduser('~/.ansible/plugins/'), '/usr/share/ansible_plugins/']])
-
```
Comment (mscherer) [inline lib/ansible/constants.py:91]: Likely unrelated cleaning, I think

File: lib/ansible/constants.py [hunk 6]
```diff
@@ -188,6 +195,7 @@ def get_plugin_paths(path):
 DEFAULT_PASSWORD_CHARS = ascii_letters + digits + ".,:-_"
 
 # non-configurable things
+DEFAULT_BECOME_PASS       = None
 DEFAULT_SUDO_PASS         = None
```
Comment (mscherer) [inline lib/ansible/constants.py:201]: Shouldn't it be removed ? ( DEFAULT_SUDO_PASS )

File: lib/ansible/playbook/__init__.py [hunk 7]
```diff
@@ -77,11 +74,13 @@ def __init__(self,
...
-        su               = False,
-        su_user          = False,
-        su_pass          = False,
         vault_password   = False,
         force_handlers   = False,
+        # privelege escalation
+        become           = False,
+        become_method    = C.DEFAULT_BECOME_METHOD,
+        become_user      = C.DEFAULT_BECOME_USER,
+        become_pass      = None,
```
Comment (mscherer) [inline lib/ansible/playbook/__init__.py]: Shouldn't it use C.DEFAULT_BECOME_PASS ?

File: lib/ansible/playbook/__init__.py [hunk 8]
```diff
@@ -92,13 +91,11 @@ def __init__(self,
...
-        sudo_pass:        if sudo==True, and a password is required, this is the sudo password
         remote_port:      default remote port to use if not specified with the host or play
         transport:        how to connect to hosts that don't specify a transport (local, paramiko, etc)
         callbacks         output callbacks for the playbook
         runner_callbacks: more callbacks, this time for the runner API
         stats:            holds aggregrate data about events occurring to each host
-        sudo:             if not specified per play, requests all plays use sudo mode
         inventory:        can be specified instead of host_list to use a pre-existing inventory object
         check:            don't change anything, just try to detect some potential changes
         any_errors_fatal: terminate the entire execution immediately when one of the hosts has failed
```
Comment (mscherer) [inline lib/ansible/playbook/__init__.py:101]: This should document self.become and others, to be consistent, I think ?

File: lib/ansible/playbook/play.py [hunk 9]
```diff
@@ -149,21 +148,32 @@ def __init__(self, playbook, ds, basedir, vault_password=None):
...
+        self.become           = ds.get('become', self.playbook.become) or self.sudo or self.su
+        self.become_method    = ds.get('become_method', self.playbook.become_method)
+        self.become_user      = ds.get('become_user', self.playbook.become_user)
+
         # gather_facts is not a simple boolean, as None means  that a 'smart'
         # fact gathering mode will be used, so we need to be careful here as
         # calling utils.boolean(None) returns False
         self.gather_facts = ds.get('gather_facts', None)
-        if self.gather_facts:
+        if self.gather_facts is not None:
```
Comment (mscherer) [inline lib/ansible/playbook/play.py:182]: Unrelated fix, I think ?

File: lib/ansible/runner/connection_plugins/zone.py [hunk 10]
```diff
@@ -68,6 +68,7 @@ def __init__(self, runner, host, port, *args, **kwargs):
         self.runner = runner
         self.host = host
         self.has_pipelining = False
+        self.SUPPORTED=['sudo', 'su', 'pbrun', 'pfexec']
```
Comment (mscherer) [inline lib/ansible/runner/connection_plugins/zone.py]: Since this module do not implement itself any of these method, shouldn't this use the constants.BECOME_METHODS ? ( since that's basically what would be implemented by ansible, so this will automatically make a new method usable by this module )

File: lib/ansible/playbook/play.py (hunk 9 above)
Comment (bcoca) [inline lib/ansible/playbook/play.py:182]: yes, I think i snuck in a couple.

File: lib/ansible/playbook/__init__.py (hunk 8 above)
Comment (bcoca) [inline lib/ansible/playbook/__init__.py:101]: the whole list is pretty out of date, almost tempted to leave as is until v2 takes over

File: lib/ansible/constants.py (hunk 6 above)
Comment (bcoca) [inline lib/ansible/constants.py:201]: actually both should

File: lib/ansible/playbook/__init__.py (hunk 7 above)
Comment (bcoca) [inline lib/ansible/playbook/__init__.py]: no, its None anyways, better just removing it

File: lib/ansible/playbook/play.py [hunk 11]
```diff
@@ -58,7 +59,7 @@ def __init__(self, playbook, ds, basedir, vault_password=None):
 
         for x in ds.keys():
             if not x in Play.VALID_KEYS:
-                raise errors.AnsibleError("%s is not a legal parameter at this level in an Ansible Playbook" % x)
+                raise errors.AnsibleError("%s is not a legal parameter of an Ansible Play" % x)
```
Comment (ghost) [inline lib/ansible/playbook/play.py:62]: Ansible Play? You mean Playbook?

File: lib/ansible/runner/__init__.py [hunk 12]
```diff
@@ -151,14 +148,15 @@ def __init__(self,
...
         accelerate_port=None,               # port to use with accelerated connection
-        su=False,                           # Are we running our command via su?
-        su_user=None,                       # User to su to when running command, ex: 'root'
-        su_pass=C.DEFAULT_SU_PASS,
         vault_pass=None,
         run_hosts=None,                     # an optional list of pre-calculated hosts to run on
         no_log=False,                       # option to enable/disable logging for a given task
         run_once=False,                     # option to enable/disable host bypass loop for a given task
-        sudo_exe=C.DEFAULT_SUDO_EXE,        # ex: /usr/local/bin/sudo
+        become=False,                         # whether to run privelege escalation or not
```
Comment (ghost) [inline lib/ansible/runner/__init__.py:155]: spelling: privilege

File: lib/ansible/playbook/play.py (hunk 11 above)
Comment (bcoca) [inline lib/ansible/playbook/play.py:62]: No, its Play, a playbook itself has no keys or anything else than plays, plays have keys and paramters.

File: lib/ansible/runner/connection_plugins/funcd.py [hunk 13]
```diff
@@ -53,16 +53,14 @@ def connect(self, port=None):
...
                      executable='/bin/sh', in_data=None, su=None, su_user=None):
         ''' run a command on the remote minion '''
 
-        if su or su_user:
-            raise errors.AnsibleError("Internal Error: this module does not support running commands via su")
-
         if in_data:
             raise errors.AnsibleError("Internal Error: this module does not support optimized module pipelining")
 
+        # totally ignores privlege escalation
```
Comment (ghost) [inline lib/ansible/runner/connection_plugins/funcd.py:63]: spelling: privilege

File: lib/ansible/runner/connection_plugins/libvirt_lxc.py [hunk 14]
```diff
@@ -65,16 +67,16 @@ def _generate_cmd(self, executable, cmd):
...
-        if su or su_user:
-            raise errors.AnsibleError("Internal Error: this module does not support running commands via su")
+        if sudoable and self.runner.become and self.runner.become_method not in self.SUPPORTED:
+            raise errors.AnsibleError("Internal Error: this module does not support running commands via %s" % self.runner.become_method)
 
         if in_data:
             raise errors.AnsibleError("Internal Error: this module does not support optimized module pipelining")
 
-        # We enter lxc as root so sudo stuff can be ignored
+        # We ignore privelege escalation!
```
Comment (ghost) [inline lib/ansible/runner/connection_plugins/libvirt_lxc.py:79]: spelling: privilege

File: lib/ansible/runner/__init__.py (hunk 12 above)
Comment (bcoca) [inline lib/ansible/runner/__init__.py:155]: he, i fixed all the misspellings in docs and methods ... but not in comments.

File: lib/ansible/runner/connection_plugins/zone.py [hunk 15]
```diff
@@ -98,17 +100,16 @@ def _generate_cmd(self, executable, cmd):
...
-        if su or su_user:
-            raise errors.AnsibleError("Internal Error: this module does not support running commands via su")
+        if sudoable and self.runner.become and self.runner.become_method not in self.SUPPORTED:
+            raise errors.AnsibleError("Internal Error: this module does not support running commands via %s" % self.runner.become_method)
 
         if in_data:
             raise errors.AnsibleError("Internal Error: this module does not support optimized module pipelining")
 
-        # We enter zone as root so sudo stuff can be ignored
+        # We happily ignore privelege escalation
```
Comment (ghost) [inline lib/ansible/runner/connection_plugins/zone.py:112]: spelling: privilege

File: lib/ansible/playbook/play.py (hunk 11 above)
Comment (ghost) [inline lib/ansible/playbook/play.py:62]: My mistake; didn't fully understand the context and previous line referenced "Playbook". Thanks for the clarification.

Comment (bcoca) [inline lib/ansible/playbook/play.py:62]: it is an unrelated change, but i opportunistically sneaked  the correction in.

File: lib/ansible/playbook/play.py [hunk 16]
```diff
@@ -149,21 +148,32 @@ def __init__(self, playbook, ds, basedir, vault_password=None):
...
-        self.su_user          = ds.get('su_user', self.playbook.su_user)
         self.no_log           = utils.boolean(ds.get('no_log', 'false'))
 
+        # TODO: make conditional on become?
+        self.sudo             = ds.get('sudo', C.DEFAULT_SUDO)
+        self.sudo_user        = ds.get('sudo_user', C.DEFAULT_SUDO_USER)
+        self.su               = ds.get('su', C.DEFAULT_SU)
+        self.su_user          = ds.get('su_user', C.DEFAULT_SU_USER)
+
+        self.become           = ds.get('become', self.playbook.become) or self.sudo or self.su
```
Comment (bcoca) [inline lib/ansible/playbook/play.py]: remove or sudo/su as we do check below and set all the vars there
